hs/src/LedgerState.hs

```diff
@@ -273,27 +270,39 @@ authTxin key txin (UTxO utxo') =
 -- than one witness.
 witnessed :: TxWits -> LedgerState -> Validity
 witnessed (TxWits tx wits) l =
-  if Set.size wits <= Set.size ins && all (hasWitness wits) ins
+  if neededSigners == signers && verified && notRedundant
     then Valid
-    else Invalid [InsufficientWitnesses]
+    else Invalid [IncorrectWitnesses]
```

Comment:
as written above, I think it makes sense to discern between insufficient and redundant witnesses